app/controllers/bulk_downloads_controller.rb

```diff
@@ -2,8 +2,55 @@
 class BulkDownloadsController < ApplicationController
   include BulkDownloadTypesHelper
 
+  before_action do
+    allowed_feature_required("bulk_downloads")
+  end
+
   # GET /bulk_downloads/types
   def types
     render json: BulkDownloadTypesHelper::BULK_DOWNLOAD_TYPES
   end
+
+  # POST /bulk_downloads
+  def create
+    bulk_download_params = params.permit(:download_type, sample_ids: [], params: {})
+
+    # Convert sample ids to pipeline run ids.
+    pipeline_run_ids = Sample.where(id: bulk_download_params[:sample_ids]).includes(:pipeline_runs)
+                             .select do |sample|
+                         # Check that the most recent pipeline run succeeded.
+                         pr = sample.pipeline_runs.first
+                         !pr.nil? && pr.finalized && pr.job_status == PipelineRun::STATUS_CHECKED
+                       end
+                             .map do |sample|
+      # For each sample, get the most recent pipeline run.
+      sample.pipeline_runs.first.id
+    end
+
+    # Throw an error if any sample doesn't have a valid pipeline run.
+    if pipeline_run_ids.length != bulk_download_params[:sample_ids].length
+      render json: {
+        error: "Some samples have either failed or not finished running yet. Please contact us for help.",
+      },
+             status: :unprocessable_entity
+      return
+    end
+
+    # Convert the sample ids to pipeline run ids.
+    bulk_download_params[:pipeline_run_ids] = pipeline_run_ids
+    bulk_download_params.delete(:sample_ids)
```

Comment:
Not great to change the input params. I would create a copy or pass the pipeline runs separately.